core/src/main/java/org/apache/iceberg/io/RollingFileWriter.java

```diff
@@ -0,0 +1,140 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A rolling writer capable of splitting incoming data/deletes into multiple files within one spec/partition.
+ */
+abstract class RollingFileWriter<T, W extends FileWriter<T, R>, R> implements FileWriter<T, R> {
+  private static final int ROWS_DIVISOR = 1000;
+
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final FileFormat fileFormat;
+  private final long targetFileSizeInBytes;
+  private final PartitionSpec spec;
+  private final StructLike partition;
+
+  private EncryptedOutputFile currentFile = null;
+  private W currentWriter = null;
+  private long currentRows = 0;
+
+  private boolean closed = false;
+
+  protected RollingFileWriter(OutputFileFactory fileFactory, FileIO io, FileFormat fileFormat,
+                              long targetFileSizeInBytes, PartitionSpec spec, StructLike partition) {
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.fileFormat = fileFormat;
+    this.targetFileSizeInBytes = targetFileSizeInBytes;
+    this.spec = spec;
+    this.partition = partition;
+  }
+
+  protected abstract W newWriter(EncryptedOutputFile file);
+
+  protected abstract long length(W writer);
+
+  protected abstract void addResult(R result);
+
+  protected abstract R aggregatedResult();
+
+  protected PartitionSpec spec() {
+    return spec;
+  }
+
+  protected StructLike partition() {
+    return partition;
+  }
+
+  @Override
+  public void write(T row) throws IOException {
+    currentWriter.write(row);
+    currentRows++;
+
+    if (shouldRollToNewFile()) {
+      closeCurrent();
+      openCurrent();
+    }
+  }
+
+  public CharSequence currentPath() {
+    Preconditions.checkNotNull(currentFile, "The currentFile shouldn't be null");
+    return currentFile.encryptingOutputFile().location();
+  }
+
+  public long currentRows() {
+    return currentRows;
+  }
+
+  protected void openCurrent() {
+    if (partition == null) {
+      this.currentFile = fileFactory.newOutputFile();
+    } else {
+      this.currentFile = fileFactory.newOutputFile(partition);
```

Comment:
This should call `newOutputFile(spec, partition)`, not this one that uses the default spec.

Reply:
In fact, this version should probably be deprecated so that we can remove it once we move over to these writers.

Reply:
Good catch, I think I did this PR before we had the new method. I'll update.